### PR TITLE
feat: anchor program prototype

### LINKED ISSUE #6: consider refactoring to Anchor
I'd like to propose we consider refactoring the AOB on-chain program to use Anchor:
- AOB already sort of uses it's own Anchor-esque macro generation (see `bonfida_utils::InstructionAccount`)
  - But it doesn't actually "give you types", as everything is still just an `AccountInfo` and you have to manually `impl` how to parse the `accounts_iter`
- A few benefits of Anchor
  - "Real types" when passing around accounts
  - Declarative specification of PDAs, account initialization, checks, etc. (separate from orderbook processor logic)
  - IDL (makes writing JS clients a lot easier and less boilerplatey)
  - Toolchain (`anchor test` is quite a nice experience, feels just like browser hot-reloading) 
  - Eliminate a lot of boilerplate and reduce size of the codebase

Downside is Anchor will require use to think about the discriminant when writing client code that does not use Anchor's built-in client. And as with any refactor we will have to test to make sure it didn't break anything (but the test coverage and usage is arguably not that great to begin with)

Please comment, especially if you don't think this is a good idea

### PR DESCRIPTION
**Pull request by lidatong**
Closes #6 

# Summary

I initialized an entire anchor project with a subdir (vs. mixing it with existing workspace). This makes it easier to diff and compare with the original. It could make sense to properly restructure the project down the line (and delete unused non-Anchor program code), but think it's better to keep both versions for now.

# Highlights

## Zero-copy
Currently using Borsh not bytemuck. Planning to refactor to Anchor zero copy (have done it separate prototype)

## Security
Anchor builds in some checks and security (which happens to also nicely abstract & separate it from the processing logic). Non-exhaustive list:

- [x] check owner handled by `Owner` trait
    - https://docs.rs/anchor-lang/0.21.0/anchor_lang/accounts/account/struct.Account.html
- [x] check uninitialized handled by `#[account(init)]`
- [x] 8 byte discriminant
- [ ] use PDAs? Anchor makes it easy with the macro (e.g. for a given market, PDAs are event_queue, bids, asks)
- [ ] rename account infos as dangerous (Anchor's `UncheckedAccount` alias for all vanilla `AccountInfo`s)

Please comment if there are other security things I missed.

# Todos
- [x] Implement the rest of the instructions
- [ ] Smoke test instructions using Anchor client and/or Solarium?
- [ ] Add proper types for event queue, bid, ask
    - `AccountInfo<'info'>` is basically an escape-hatch
    - instead we want `Account<'info', T>` wherever possible to let Anchor do the work for us (serde, owner check, etc.)
- [ ] `Slab` holds a mutable reference, so it can't impl `Clone` (which Anchor traits require). So we can't easily do Anchor codegen. Perhaps consider owned slice again?
- [ ] Anchor generates it's own cargo workspace, and I can't roll it up into the top-level otherwise Anchor gives an error "not an anchor workspace"?
- [ ] Dynamically allocate size of `bids` and `asks` (hardcoded to max of `10240` right now)

### CODE REVIEW

**Comment by yuhanfang** [comment]
Adding Jarry for awareness. I'll have to learn anchor to review this... for the time being as long as these are totally independent of the non-anchor code (looks like that's true) then I'm fine merging piecemeal whenever there is a green checkpoint. 

Would you mind adding a small solarium-style smoke test for the create market instruction in your next PR?

**Comment by lidatong** [comment]
sounds good

> as long as these are totally independent
yes, true EXCEPT for one thing, which is Anchor needs to generate IDL for types, and this requires the library (of data structures) to be "Anchor-aware". this does not break anything, but there are API implications. i'm still thinking through it, and would be good to discuss with Jarry

> solarium-style smoke test
for sure, i'll check out Solarium -- haven't looked yet

btw Anchor actually includes their own [JS testing kit](https://github.com/foonetic/agnostic-orderbook/pull/10/files#diff-01453dc6b4f665c83943e0208f678c35afd1c9823a8603414839c8d986e69cfb) with a lot of codegen. `anchor test` is a one-click button to build, deploy, run tests all at once (using the IDL to do codegen / reduce boilerplate)

**Comment by lidatong** [comment]
> yes, true EXCEPT for one thing, which is Anchor needs to generate IDL for types, and this requires the data structure / library crate to be "Anchor-aware"

I've been thinking about this for a while, and here is my brain dump. It's an interesting API design question.

**First to clarify what I'm even talking about:**
- Imagine you have a type `T`: you want to tell Anchor "hey this type `T` corresponds to this `AccountInfo`". In order to "plug in" to Anchor, you have to implement some traits Anchor defines (e.g. `AnchorSerialize`).
- Normally Anchor just codegens this all for you when you attach a `derive` macro. But what if you can't and don't want to modify that type `T`!

**Here is a concrete example from Anchor's own codebase. They wanted to implement the Anchor traits for the types on SPL Token program (which existed before Anchor).**
- You can't just change modify SPL Token source code to derive Anchor traits (_well, I guess they could've since Solana team knows Anchor team_)
- Essentially their pattern was to create a wrapper type, and derive the Anchor traits on the struct wrapper.
  - https://github.com/project-serum/anchor/blob/master/spl/src/token.rs#L306
- It's important to take note of Rust's orphan rule. **_Normally, what Anchor did above wouldn't be allowed._** You cannot implement a trait from one 3rd-party crate for another 3rd party crate's type `T`.
  - https://doc.rust-lang.org/book/ch10-02-traits.html
- So I wouldn't have been able to implement the Anchor traits for SPL Token in my own separate "orderbook" crate, for example. By virtue of orphan rule, that code _must_ live inside either the Anchor source code or SPL Token source code.

**We're in a very similar situation here:**
- We want to implement Anchor traits for our orderbook data structure. This makes it so Anchor manages the `serde`, `owner`, `discriminator`, etc. checks for us (saving us both boilerplate and giving us more security).
- We also get "types": it's not just an opaque `AccountInfo`, but Anchor codegens a real wrapper type for us that knows about the struct with all public methods, fields, etc.

**So here's the question:**
- Our situation is interesting because we could just attach a macro to the data structure crate itself.
- Or we could take a "wrapper" type approach, where inside our Anchor program

This might seem a lot of text over a relatively minor thing, but I think this API design question could have quite large implications: **_broadly, how do we present our _interface_ to users / callers?_**

Our hands might be tied in the future to make changes, in order to preserve API compatibility (though we should also think about API evolution).

**Comment by lidatong** [comment]
Turns out you actually lose even more than I thought in ergonomics without `#[account]`... specifically, if you take the "wrapper type" approach, this type is unknown to the IDL and you end up having to do all the manual fudging around with bytes / serdes Anchor was supposed to take care of for you on the client side

File: anchor/programs/anchor-agnostic-orderbook/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+    pub asks: AccountInfo<'info>,
+    #[account(mut)]
+    pub payer: Signer<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+pub struct NewOrder<'info> {
+    #[account(mut)]
+    pub market: Account<'info, MarketState>,
```
Comment (jarry-xiao) [inline anchor/programs/anchor-agnostic-orderbook/src/lib.rs:331]: I think it's definitely possible to make the `MarketState` zero-copy by using `AccountLoader`

**Review by jarry-xiao — APPROVED** [review]
Given that all of this stuff is experimental, I wouldn't be too afraid of breaking existing interfaces to help discover better implementations. The following would be good to have measured:
```
Configuration Type (Serum V3, AOB, Anchor AOB, etc.)
test_name_0, avg_insert_time_0, avg_delete_time_0
test_name_1, avg_insert_time_1, avg_delete_time_1
...
test_name_N, avg_insert_time_N, avi_delete_time_N
```
Then based on each test, we can weigh the tradeoffs of each implementation. I think it should even be possible to put an enum flag on the market state to determine which tree/matching engine is used for a given market, and given the benchmarks, the downstream user/application will have an understanding of the features/limitations of a particular deployed market.
Getting into anchor is a good first step because I think as we run into issues with loading and serializing these types of dynamic objects (orderbooks/event queues), Armani and co can figure out clean ways to handle this more seamlessly -> more complex + efficient data structure support throughout all solana programs without having to chew the glass of manually computing buffer offsets.

**Comment by lidatong** [comment]
Thanks for the review @jarry-xiao!

> Given that all of this stuff is experimental, I wouldn't be too afraid of breaking existing interfaces to help discover better implementations.
yep makes sense. what are your thoughts on "separation of concerns" perspective, e.g. data structure should be Anchor-unaware vs. having this `#[account]` thing. i don't have that strong feelings, esp. if Anchor becomes the de-facto standard

> put an enum flag on the market state
in the same program? what do you think of different deployed programs instantiating different orderbooks?

> Getting into anchor is a good first step because I think as we run into issues with loading and serializing
absolutely agree. Anchor gives a lot + the abstraction

and definitely some glass chewing. you have to dig a bit through the Anchor codebase for the more uncommon use cases. would be great to see Anchor team keep making it better and adding larger examples (perhaps the AOB could even become one!)

**Comment by jarry-xiao** [comment]
> what are your thoughts on "separation of concerns" perspective, e.g. data structure should be Anchor-unaware vs. having this #[account] thing.

I'm in the same boat here, could go either way. I think the more anchor we use, the more potentially help we can leverage from the anchor community.

> in the same program? what do you think of different deployed programs instantiating different orderbooks?

This might be a cleaner separation that allows for more backwards compatibility. I think at some point it might be good to have a whitelist of approved program ids with audited orderbook contracts that comply with some universal interface

**Comment by lidatong** [comment]
> the more potential help we can leverage from the anchor community
Yes I definitely ran into a lot of things, and while the anchor discord is helpful my questions went unanswered

> whitelist of approved program ids with audited orderbook contracts
Absolutely

**Pull request merged**